Return a non-zero exit code (or a flag for that option) when files fail to update on `hs upload` command

## Description and Context
<!-- Provide a summary of the issue -->
<!-- Provide links to relevant information to promote understanding and addressing this issue -->
<!-- Which version of the CLI you are using? -->

Could a non-zero exit code be returned when a file doesn't upload on the `hs upload` command due to an error? We run this command with a GitHub deploy action and saw that the deployment was showing to be successful even though some module files didn't update due to errors.


## Expected behavior
<!-- What should the behavior be instead? -->
I would expect to receive a non-zero exit code to alert that there was an error with the upload, as of right now you won't know until you notice some of your modules didn't update.

## Screenshots
<!-- Provide an image of the observed/expected behavior -->
![Screen Shot 2022-01-07 at 10 03 05 AM](https://user-images.githubusercontent.com/65314061/148572546-7a2bea23-1fcd-43ea-9085-029dbf3d29b6.png)

Ok, I am **not** a javascript developer, but I tried and failed in the first step.
**Simply adding a process.exit did not work.**

To reproduce the error case, I did the following:
```
hs create webpack-serverless
```
This command created a folder with the webpack-serverless boilerplate code.

Then I tried to upload this folder with
```
hs upload webpack-serverless webpack-serverless
```
This fails with
```
[ERROR] Uploading file "..."
[ERROR] Failed to read the source for the serverless config file
```


**Second attempt**
I took a closer look ath the "uploadFolder" method in uploadFolder.js and came across another catch(error) block in line 87 with error handling and logging. The problem is that the error is not passed on to the parent method in all cases.

After a little fix I got this :-)

![image](https://user-images.githubusercontent.com/57138594/154794316-3597744a-ac8f-4f8f-ae95-f9048081c8cb.png)

--- a/packages/cli-lib/index.js
+++ b/packages/cli-lib/index.js
@@ -16,7 +16,7 @@ const {
   writeConfig,
 } = require('./lib/config');
 const { checkAndWarnGitInclusion } = require('./lib/git');
-const { uploadFolder } = require('./lib/uploadFolder');
+const { hasUploadErrors, uploadFolder } = require('./lib/uploadFolder');
 const { watch } = require('./lib/watch');
 const { read } = require('./lib/read');
 const { walk } = require('./lib/walk');
@@ -37,6 +37,7 @@ module.exports = {
   getAccountId,
   getPortalConfig: getAccountConfig,
   getPortalId: getAccountId,
+  hasUploadErrors,
   updateAccountConfig,
   updatePortalConfig: updateAccountConfig,
   uploadFolder,

--- a/packages/cli-lib/lang/en.lyaml
+++ b/packages/cli-lib/lang/en.lyaml
@@ -614,6 +614,7 @@ en:
           invalidExtension: "The file \"{{ path }}\" does not have a valid extension"
           invalidPath: "The path \"{{ path }}\" is not a path to a file or folder"
           uploadFailed: "Uploading file \"{{ src }}\" to \"{{ dest }}\" failed"
+          someFilesFailed: "One or more files failed to upload to \"{{ dest }}\" in the Design Manager"
         positionals:
           src:
             describe: "Path to the local file, relative to your current working directory."

--- a/packages/cli-lib/lib/uploadFolder.js
+++ b/packages/cli-lib/lib/uploadFolder.js
@@ -155,9 +155,8 @@ async function uploadFolder(accountId, src, dest, options) {
 }
 
 function hasUploadErrors(results) {
-  return (
-    results.filter(result => result.resultType === FileUploadResultType.FAILURE)
-      .length > 0
+  return results.some(
+    result => result.resultType === FileUploadResultType.FAILURE
   );
 }
 

--- a/packages/cli-lib/lib/uploadFolder.js
+++ b/packages/cli-lib/lib/uploadFolder.js
@@ -19,6 +19,11 @@ const {
   isFatalError,
 } = require('../errorHandlers');
 
+const FileUploadResultType = {
+  SUCCESS: 'SUCCESS',
+  FAILURE: 'FAILURE',
+};
+
 const queue = new PQueue({
   concurrency: 10,
 });
@@ -112,13 +117,18 @@ async function uploadFolder(accountId, src, dest, options) {
     await queue.addAll(filesToUpload.map(uploadFile));
   }
 
-  return queue.addAll(
+  const results = await queue.addAll(
     failures.map(({ file, destPath }) => {
       return async () => {
         logger.debug('Retrying to upload file "%s" to "%s"', file, destPath);
         try {
           await upload(accountId, file, destPath, apiOptions);
           logger.log('Uploaded file "%s" to "%s"', file, destPath);
+          return {
+            resultType: FileUploadResultType.SUCCESS,
+            error: null,
+            file,
+          };
         } catch (error) {
           logger.error('Uploading file "%s" to "%s" failed', file, destPath);
           if (isFatalError(error)) {
@@ -132,12 +142,27 @@ async function uploadFolder(accountId, src, dest, options) {
               payload: file,
             })
           );
+          return {
+            resultType: FileUploadResultType.FAILURE,
+            error,
+            file,
+          };
         }
       };
     })
   );
+  return results;
+}
+
+function hasUploadErrors(results) {
+  return (
+    results.filter(result => result.resultType === FileUploadResultType.FAILURE)
+      .length > 0
+  );
 }
 
 module.exports = {
+  hasUploadErrors,
+  FileUploadResultType,
   uploadFolder,
 };

--- a/packages/cli/commands/upload.js
+++ b/packages/cli/commands/upload.js
@@ -1,7 +1,7 @@
 const fs = require('fs');
 const path = require('path');
 
-const { uploadFolder } = require('@hubspot/cli-lib');
+const { uploadFolder, hasUploadErrors } = require('@hubspot/cli-lib');
 const { getFileMapperQueryValues } = require('@hubspot/cli-lib/fileMapper');
 const { upload } = require('@hubspot/cli-lib/api/fileMapper');
 const {
@@ -163,13 +163,22 @@ exports.handler = async options => {
     uploadFolder(accountId, absoluteSrcPath, dest, {
       mode,
     })
-      .then(() => {
-        logger.success(
-          i18n(`${i18nKey}.success.uploadComplete`, {
-            dest,
-          })
-        );
-        logThemePreview(src, accountId);
+      .then(results => {
+        if (!hasUploadErrors(results)) {
+          logger.success(
+            i18n(`${i18nKey}.success.uploadComplete`, {
+              dest,
+            })
+          );
+          logThemePreview(src, accountId);
+        } else {
+          logger.error(
+            i18n(`${i18nKey}.errors.someFilesFailed`, {
+              dest,
+            })
+          );
+          process.exit(EXIT_CODES.ERROR);
+        }
       })
       .catch(error => {
         logger.error(

--- a/packages/cli/commands/upload.js
+++ b/packages/cli/commands/upload.js
@@ -150,6 +150,7 @@ exports.handler = async options => {
             payload: src,
           })
         );
+        process.exit(EXIT_CODES.ERROR);
       });
   } else {
     logger.log(
@@ -180,6 +181,7 @@ exports.handler = async options => {
         logErrorInstance(error, {
           accountId,
         });
+        process.exit(EXIT_CODES.ERROR);
       });
   }
 };

--- a/packages/cli/commands/upload.js
+++ b/packages/cli/commands/upload.js
@@ -55,7 +55,7 @@ exports.handler = async options => {
   await loadAndValidateOptions(options);
 
   if (!validateMode(options)) {
-    process.exit(EXIT_CODES.ERROR);
+    process.exit(EXIT_CODES.WARNING);
   }
 
   const accountId = getAccountId(options);
@@ -98,7 +98,7 @@ exports.handler = async options => {
 
   if (srcDestIssues.length) {
     srcDestIssues.forEach(({ message }) => logger.error(message));
-    process.exit(EXIT_CODES.ERROR);
+    process.exit(EXIT_CODES.WARNING);
   }
   if (stats.isFile()) {
     if (!isAllowedExtension(src)) {
@@ -150,7 +150,7 @@ exports.handler = async options => {
             payload: src,
           })
         );
-        process.exit(EXIT_CODES.ERROR);
+        process.exit(EXIT_CODES.WARNING);
       });
   } else {
     logger.log(
@@ -177,7 +177,7 @@ exports.handler = async options => {
               dest,
             })
           );
-          process.exit(EXIT_CODES.ERROR);
+          process.exit(EXIT_CODES.WARNING);
         }
       })
       .catch(error => {
@@ -190,7 +190,7 @@ exports.handler = async options => {
         logErrorInstance(error, {
           accountId,
         });
-        process.exit(EXIT_CODES.ERROR);
+        process.exit(EXIT_CODES.WARNING);
       });
   }
 };

